docs: Routes are wrapped w/ Activity in Cache Components (#85309)
vercel/next.js: docs/01-app/01-getting-started/04-linking-and-navigating.mdx

- Using the [`@next/bundle-analyzer`](/docs/app/guides/package-bundling#analyzing-javascript-bundles) plugin to identify and reduce bundle size by removing large dependencies.
- Moving logic from the client to the server where possible. See the [Server and Client Components](/docs/app/getting-started/server-and-client-components) docs for guidance.

## Navigation with Cache Components

When the [`cacheComponents`](/docs/app/api-reference/config/next-config-js/cacheComponents) flag is enabled, Next.js uses React's [`<Activity>`](https://react.dev/reference/react/Activity) component to preserve component state during client-side navigation.

> Instead of unmounting the previous route when you navigate away, Next.js hides it with `display: none`. This means:

Author: This is an impl detail, although true, this is done by React when we setting Activity to hidden

- Component state is preserved when navigating between routes
- When you navigate back, the previous route reappears with its state intact
- Effects are cleaned up when a route is hidden, and recreated when it becomes visible again

This behavior improves the navigation experience by maintaining UI state (form inputs, or expanded sections) when users navigate back and forth between routes.

> **Good to know**: Next.js uses heuristics to keep a few recently visited routes hidden with Activity to preserve their state, while older routes are removed from the DOM to prevent excessive growth.

<details>
<summary>**Portals under Activity require manual cleanup**</summary>

> As a consequence of routes being hidden rather than unmounted, portals created via [`createPortal`](https://react.dev/reference/react-dom/createPortal) need special handling. Since portals render outside their parent component's DOM tree, they bypass the CSS cascade and remain visible even when the parent route is hidden.

Author: Let's link to React's docs here, showing what the hidden mode does

React can't automatically hide portals because it doesn't control the portal container—your effect does. To fix this, add a cleanup function to your effect:

```tsx highlight={6-10}
useLayoutEffect(() => {
  const el = document.getElementById('portal-root')
  setContainer(el)

  // Hide the portal container when this component is hidden by Activity
  return () => {
    if (el) el.style.display = 'none'
  }
}, [])
```

The cleanup function runs when the component's Effects are cleaned up by the Activity boundary.

> Whether portal elements should automatically become visible again when the Activity is shown depends on your specific UI needs.

Author: `...when the Activity is visible again...`

- Moving logic from the client to the server where possible. See the [Server and Client Components](/docs/app/getting-started/server-and-client-components) docs for guidance.

## Navigation with Cache Components

When the [`cacheComponents`](/docs/app/api-reference/config/next-config-js/cacheComponents) flag is enabled, Next.js uses React's [`<Activity>`](https://react.dev/reference/react/Activity) component to preserve component state during client-side navigation.

Rather than unmounting the previous route when you navigate away, Next.js sets the Activity mode to [`"hidden"`](https://react.dev/reference/react/Activity#activity). This means:

- Component state is preserved when navigating between routes
- When you navigate back, the previous route reappears with its state intact
- Effects are cleaned up when a route is hidden, and recreated when it becomes visible again

This behavior improves the navigation experience by maintaining UI state (form inputs, or expanded sections) when users navigate back and forth between routes.

> > **Good to know**: Next.js uses heuristics to keep a few recently visited routes hidden with Activity to preserve their state, while older routes are removed from the DOM to prevent excessive growth.

Author: ```suggestion
> **Good to know**: Next.js uses heuristics to keep a few recently visited routes [`"hidden"`](https://react.dev/reference/react/Activity#activity), while older routes are removed from the DOM to prevent excessive growth.
```

<details>
<summary>**Portals under Activity require manual cleanup**</summary>

As a consequence of routes being hidden rather than unmounted, portals created via [`createPortal`](https://react.dev/reference/react-dom/createPortal) need special handling. Since portals render outside their parent component's DOM tree, they remain visible even when the parent route is hidden.

React can't automatically hide portals because it doesn't control the portal container, your effect does. To fix this, add a cleanup function to your effect:

```tsx highlight={6-10}
useLayoutEffect(() => {
  const el = document.getElementById('portal-root')
  setContainer(el)

  // Hide the portal container when this component is hidden by Activity
  return () => {
    if (el) el.style.display = 'none'
  }

> }, [])

Reviewer: I'm not sure this works for all cases. I've tried this solution and it doesn't seem to work.

Plus, is this being proposed as a workaround or as an actual solution? I think this should be handled at the framework side since users don't expect Next.js to hide their previous route with `Activity`

I think Next.js should either hide portals on navigation or revert #84923 and make `routerBfCache` optional again, not attached to `cacheComponents`

For example, I'd like to use `cacheComponents` but not `bfCache` until this behavior is fixed. They shouldn't be under the same flag

Author: > I'm not sure this works for all cases. I've tried this solution and it doesn't seem to work.

I tried a few variations on my React codesandbox and it worked though. https://codesandbox.io/p/sandbox/react-dev-forked-626mp3?file=%2Fsrc%2FSidebar.js

But I can imagine in some cases the code controlling the container might be handled by UI library code, and not the developer code necessarily.

> Plus, is this being proposed as a workaround or as an actual solution? I think this should be handled at the framework side since users don't expect Next.js to hide their previous route with Activity
> 
> I think Next.js should either hide portals on navigation

This is a React Activity + Portals limitation. Next.js might separate the config option, but the root issue will still be there. 

Part of the reason I haven't merged is to collect this kind of feedback, but also because today we'll have time to discuss this. I'll try to post an update when I know more.

Reviewer: Thanks for looking into this and sharing the codesandbox example.

I agree this should also be looked into on React side, so I’ve also opened an issue upstream to track this:
https://github.com/facebook/react/issues/35000

That being said, I'm not sure this is a bug or expected behavior for the `Activity` component.

Manually hiding the portal container can work when we own the code that manages the portal target.

The challenge in this case is that developers aren’t explicitly choosing to wrap routes in `Activity`. Next.js now applies it automatically when `cacheComponents` is enabled, and there’s currently no way to opt out of the hidden-routes behavior without also disabling component caching entirely.

If I were using `Activity` directly in my code, I could choose not to combine it with portals or implement a visibility-aware container. But since this is happening inside the framework, we don’t have control over the portal target or lifecycle, and some libraries manage their own portal containers. This makes the workaround impractical or impossible for many real-world setups.

There are also other side-effect concerns documented in React’s Activity docs. Some apps may want component caching but not hidden-route behavior exactly because of these trade-offs.

That’s why I believe `cacheComponents` and the `routerbfCache` should ideally be separate flags, so users can adopt one without implicitly opting into the other.

Happy to help test any changes on either side. Thanks again for taking the time to discuss before merging.

Author: I'll move the note about Activity being used in Routes to new Cache Components entry, it makes sense to be there, there's already a lot of info on Linking and Navigating, and this only happens under Cache Components. 

Regarding the portals, I don't know yet, so I'll will remove its mention - this is not a `Closes: XYZ` PR either, and we need to mention that Activity wraps routes in Cache Components. It does feel like it is more in React's field at the moment. Thanks for opening an issue over there.
